### PR TITLE
feat(storage):  improve optimize and recluster

### LINKED ISSUE #11799: Feature: allow specify a LIMIT clause in the OPTIMZE TABLE ... PURGE statement
**Summary**

On scheduling the background tasks (including reclusters, purge, vacuum, etc.), the background task scheduler would hope each task could be executed in a bounded time range.

Currently we support:

```
OPTIMIZE TABLE [database.]table_name PURGE BEFORE (TIMESTAMP => '<TIMESTAMP>'::TIMESTAMP)
```

It would be neat if we can specify a LIMIT clause together.

The query to be executed by the background task scheduler would looks like:

```
OPTIMIZE TABLE [database.]table_name PURGE BEFORE (TIMESTAMP => '1 day ago') LIMIT 100
```

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. Add limit for recluster.  

```
ALTER TABLE [IF EXISTS] <name> RECLUSTER [FINAL] [WHERE condition] [LIMIT <segment_count>]
```
The option LIMIT sets the maximum number of segments to be recluster. Databend will select the newest segments. The maximum segment_ount limit is `max_storage_io_requests`.

2. Add memory usage limit for recluster

3. Sort the blockmeta by cluster_statistics during compact

4. after optimize compact, do recluster

5. Optimize recluster, serialize block in parallel during recluster.
   
```
create table test_order (
    id bigint,
    id1 bigint,
    id2 bigint,
    id3 bigint,
    id4 bigint,
    id5 bigint,
    id6 bigint,
    id7 bigint,
    
    s1 varchar,
    s2 varchar,
    s3 varchar,
    s4 varchar,
    s5 varchar,
    s6 varchar,
    s7 varchar,
    s8 varchar,
    s9 varchar,
    s10 varchar,
    s11 varchar,
    s12 varchar,
    s13 varchar,
    
    d1 DECIMAL(20, 8),
    d2 DECIMAL(20, 8),
    d3 DECIMAL(20, 8),
    d4 DECIMAL(20, 8),
    d5 DECIMAL(20, 8),
    d6 DECIMAL(30, 8),
    d7 DECIMAL(30, 8),
    d8 DECIMAL(30, 8),
    d9 DECIMAL(30, 8),
    d10 DECIMAL(30, 8),
    
    insert_time datetime,
    insert_time1 datetime,
    insert_time2 datetime,
    insert_time3 datetime,
    
    i int
) CLUSTER BY(id,insert_time);


create table random_source (
    id bigint,
    id1 bigint,
    id2 bigint,
    id3 bigint,
    id4 bigint,
    id5 bigint,
    id6 bigint,
    id7 bigint,
    
    s1 varchar,
    s2 varchar,
    s3 varchar,
    s4 varchar,
    s5 varchar,
    s6 varchar,
    s7 varchar,
    s8 varchar,
    s9 varchar,
    s10 varchar,
    s11 varchar,
    s12 varchar,
    s13 varchar,
    
    d1 DECIMAL(20, 8),
    d2 DECIMAL(20, 8),
    d3 DECIMAL(20, 8),
    d4 DECIMAL(20, 8),
    d5 DECIMAL(20, 8),
    d6 DECIMAL(30, 8),
    d7 DECIMAL(30, 8),
    d8 DECIMAL(30, 8),
    d9 DECIMAL(30, 8),
    d10 DECIMAL(30, 8),
    
    insert_time datetime,
    insert_time1 datetime,
    insert_time2 datetime,
    insert_time3 datetime,
    
    i int
) Engine = Random;

insert into test_order select * from random_source limit 50000000;
```

Before optimize:
```
mysql> optimize table test_order compact;
Query OK, 32000000 rows affected (3 min 10.62 sec)
```

After:
```
mysql> optimize table test_order compact;
Query OK, 32000000 rows affected (55.94 sec)
```

```
The recluster Pipeline is as follows:
┌──────────┐     ┌───────────────┐     ┌─────────┐
│FuseSource├────►│CompoundBlockOp├────►│SortMerge├────┐
└──────────┘     └───────────────┘     └─────────┘    │
┌──────────┐     ┌───────────────┐     ┌─────────┐    │     ┌──────────────┐     ┌─────────┐
│FuseSource├────►│CompoundBlockOp├────►│SortMerge├────┤────►│MultiSortMerge├────►│Resize(N)├───┐
└──────────┘     └───────────────┘     └─────────┘    │     └──────────────┘     └─────────┘   │
┌──────────┐     ┌───────────────┐     ┌─────────┐    │                                        │
│FuseSource├────►│CompoundBlockOp├────►│SortMerge├────┘                                        │
└──────────┘     └───────────────┘     └─────────┘                                             │
┌──────────────────────────────────────────────────────────────────────────────────────────────┘
│         ┌──────────────┐
│    ┌───►│SerializeBlock├───┐
│    │    └──────────────┘   │
│    │    ┌──────────────┐   │    ┌─────────┐    ┌────────────────┐     ┌─────────────────┐     ┌──────────┐
└───►│───►│SerializeBlock├───┤───►│Resize(1)├───►│SerializeSegment├────►│TableMutationAggr├────►│CommitSink│
     │    └──────────────┘   │    └─────────┘    └────────────────┘     └─────────────────┘     └──────────┘
     │    ┌──────────────┐   │
     └───►│SerializeBlock├───┘
          └──────────────┘
```

Closes #11799

### CODE REVIEW

**Comment by BohuTANG** [comment]
@zhyass 
Could you add some informations about what's the [limit N] is? Like:
`opetimize table xx purge limit 100` -- what's the `limit 100` under the hood?

File: tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table [hunk 1]
```diff
@@ -557,25 +557,36 @@ query I
...
+1 1
 
 query TIIFFT
 select * from clustering_information('db_09_0008','t10')
 ----
-(abs(a)) 2 0 1.0 2.0 {"00002":2}
+(abs(a)) 2 0 0.0 1.0 {"00001":2}
+
+statement ok
+optimize table t10 purge limit 2
```
Comment (BohuTANG) [inline tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table:584]: During the test, it appears that we do not verify whether the recluster is expected to occur after `compact limit N`.

File: tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table [hunk 2]
```diff
@@ -590,6 +590,30 @@ select count() from fuse_snapshot('db_09_0008', 't10')
...
+statement ok
+insert into t11 values(1),(0),(3)
+
+query TIIFFT
+select * from clustering_information('db_09_0008','t11')
+----
+(abs(a)) 2 0 1.0 2.0 {"00002":2}
+
+statement ok
+optimize table t11 compact
```
Comment (BohuTANG) [inline tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table:608]: Should with LIMIT?

Comment (zhyass) [reply tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table:608]: > Should with LIMIT?

No, just test compact will do recluster

**Comment by ZhiHanZ** [comment]
wonder about the mechanism, does the purge still needs to store candidate in memory and delete all candidates in a old segment?

**Comment by zhyass** [comment]
> wonder about the mechanism, does the purge still needs to store candidate in memory and delete all candidates in a old segment?

will select and purge the oldest snapshots.

**Comment by BohuTANG** [comment]
## Summary(By llmchain.rs)
* **Added support for "limit" parameter in OPTIMIZE TABLE and ALTER TABLE RECLUSTER statements**
The code changes added support for the "limit" parameter in the ReclusterTable action, OptimizeTableStmt struct, and Display implementation of OptimizeTableStmt. The changes also added support for parsing the "limit" parameter in the OptimizeTableStmt and added new test cases for OPTIMIZE TABLE statement with different options and limits.

* **Changed "dry_run_limit" parameter to "dry_run" boolean parameter**
The code changes modified the `dry_run_limit` parameter to a `dry_run` boolean parameter in the `vacuum_handler.rs` and `handler.rs` files. The changes also added a constant `DRY_RUN_LIMIT` and modified the function signature of `do_vacuum` to take a boolean `dry_run` instead of an optional `dry_run_limit`.

* **Modified function signatures to take a Vec<DataBlock> instead of a slice &[DataBlock]**
The code changes modified the function signatures of `compact_final` in `transform_block_compact.rs`, `transform_block_compact_for_copy.rs`, and `transform_compact.rs` to take a `Vec<DataBlock>` instead of a slice `&[DataBlock]`. The changes also modified the `consume_event` function to check if the output port can push before pushing the next data block.

* **Added support for TransformSerializeBlock and TransformSerializeSegment**
The code changes added support for `TransformSerializeBlock` and `TransformSerializeSegment` in various files, including `transform_append.rs`, `transform_serialize_data.rs`, and `replace.rs`. The changes also removed the unused `AppendTransform` and added a new module `mutation_meta`.

* **Added limit parameter to various functions**
The code changes added a `limit` parameter to the `compact` method in `table.rs`, the `do_purge` function in `gc.rs`, and the `do_recluster` function in `recluster.rs`. The changes also modified the implementation of the `do_recluster` function to use the limit parameter to limit the number of segment locations processed at a time.

* **Changed various function parameters and return types**
The code changes changed the `dry_run_limit` parameter to a boolean `dry_run` parameter in various files, including `vacuum_handler.rs`, `handler.rs`, and `gc.rs`. The changes also changed the return type of `apply_delete` function in `merge_into_mutator.rs` and the function signature of `compact_table` function in `hive_table.rs`.

**Comment by flaneur2020** [comment]
> wonder about the mechanism, does the purge still needs to store candidate in memory and delete all candidates in a old segment?

IMO purge do not need scan the world, but only need just tailing the older snapshot & segment & block files.

If I understands correctly, the block files have some kind of time ordering, if a block file's creation time earlier than the earliest active snapshot & not included in this snapshot, then it can be safely purged.

**Comment by zhyass** [comment]
> IMO purge do not need scan the world, but only need just tailing the older snapshot & segment & block files.
> 
> If I understands correctly, the block files have some kind of time ordering, if a block file's creation time earlier than the earliest active snapshot & not included in this snapshot, then it can be safely purged.

This can be used by purge orphan blocks in vacuum table.

**Comment by zhyass** [comment]
askbend:summary

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 I applied the following changes:

- Add `limit` for vacuum, optimize table optimize, and recluster
- Separate `TransformSerializeData` to `TransformSerializeBlock` and `TransformSerializeSegment`. The former serializes a data block, the latter serializes a segment.
- Add statistical info to `ReclusterMutator`, and return the blocks that need to be deleted.
- Add a memory threshold for `ReclusterMutator`
- Handle `PipelineBuilder` in a more generic way, by using `TransformSerializeBlock` and `TransformSerializeSegment`
- Add test case for `optimize table compact limit`

Does this look ok? Let me know if you have any questions or feedback.

**Comment by BohuTANG** [comment]
> ### PR summary(by Claude):
> I applied the following changes:
> 
> * Add `limit` for vacuum, optimize table optimize, and recluster
> * Separate `TransformSerializeData` to `TransformSerializeBlock` and `TransformSerializeSegment`. The former serializes a data block, the latter serializes a segment.
> * Add statistical info to `ReclusterMutator`, and return the blocks that need to be deleted.
> * Add a memory threshold for `ReclusterMutator`
> * Handle `PipelineBuilder` in a more generic way, by using `TransformSerializeBlock` and `TransformSerializeSegment`
> * Add test case for `optimize table compact limit`
> 
> Does this look ok? Let me know if you have any questions or feedback.

Looks good, what's your prompt?

**Comment by drmingdrmer** [comment]
> Looks good, what's your prompt?

quite simple, just:

```bash
pr=$1

prompt="$(
echo 'write human readable git commit message for the following patch, in detail';
echo '```diff';
gh pr diff $1;
echo '```';
)"
```

The response from Claude is inconsistent, with only the first one being a reasonable summary. It is unclear why subsequent prompts of the same content result in messy replies. I'll fire several other summarization and you'll see.

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 This is git commit message for the following patch:

```
diff --git a/Cargo.lock b/Cargo.lock
index 015b4adb5a3..04f369e7ae4 100644

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 I applied your suggestions

> diff --git a/src/query/service/src/interpreters/interpreter_table_optimize.rs b/src/query/service/src/interpreters/interpreter_table_optimize.rs
> index 22dbcdb4c48..8db7a9aca15 100644
>

**Comment by drmingdrmer** [comment]
> ### PR summary(by Claude):
> This is git commit message for the following patch:
> 
> ```
> diff --git a/Cargo.lock b/Cargo.lock
> index 015b4adb5a3..04f369e7ae4 100644
> ```

Claude has lost its mind. @BohuTANG

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 I suggest the following summary for these changes:

`Cargo.lock`:
- Added `sys-info` crate as a dependency 

`src/query/ast/src/ast/format/ast_format.rs`:
- Added support for LIMIT clause in RECLUSTER TABLE statement

`src/query/ast/src/ast/format/syntax/ddl.rs`:
- Added pretty printing for LIMIT clause in ALTER TABLE RECLUSTER statement

`src/query/ast/src/ast/statements/table.rs`:
- Added `limit` field to `ReclusterTable` and `OptimizeTable` enums to support LIMIT clause

`src/query/ast/src/parser/statement.rs`:
- Added parsing for LIMIT clause in RECLUSTER TABLE and OPTIMIZE TABLE statements

`src/query/ast/tests/it/parser.rs`: 
- Added test cases for LIMIT clause in RECLUSTER TABLE and OPTIMIZE TABLE statements

`src/query/catalog/src/table.rs`:
- Added `limit` parameter to `recluster()` and `purge()` table methods

`src/query/ee-features/vacuum-handler/src/vacuum_handler.rs`:
- Renamed `dry_run_limit` parameter to `dry_run`

`src/query/ee/src/storages/fuse/operations/handler.rs`:
- Reflect the change in API 

`src/query/ee/src/storages/fuse/operations/vacuum.rs`:
- Extracted `DRY_RUN_LIMIT` constant and use `dry_run` flag instead of `dry_run_limit`  

`src/query/pipeline/transforms/src/*`:
- Various refactors and changes to compact transforms

`src/query/service/src/interpreters/*`:
- Interpreters for OPTIMIZE TABLE and RECLUSTER TABLE now accept LIMIT clauses and pass them to respective table methods.

Hope this summary covers the major changes and let me know if you need any clarifications!

**Comment by BohuTANG** [comment]
> > Looks good, what's your prompt?
> 
> quite simple, just:

Got. Claude has a large context than OpenAI, in OpenAI, we must split this diff into parts, let openai explain each part with this prompt:
https://github.com/shafishlabs/llmchain.rs/blob/8e04987f3264508a9ddd4f921356a8f60a12aff5/llmchain/memory/src/github_pr_summary.rs#L52-L63

Then let the openai give a final summary based on the part summary with prompt:
https://github.com/shafishlabs/llmchain.rs/blob/8e04987f3264508a9ddd4f921356a8f60a12aff5/llmchain/prompts/src/github_pr_summary_prompt.rs#L33-L40

**Comment by drmingdrmer** [comment]
### PR summary(by Claude):

 This is a summary of the main changes made in the pull request:

* Added limit parameter to optimize table, purge and recluster operations 

   - optimize table compact now supports a limit parameter to specify the maximum number of blocks to compact. This is useful for controlling the impact on the cluster.

   - purge now supports a limit parameter to specify the maximum number of snapshot files to purge. This avoids purging too many files at once.  

   - recluster now supports a limit parameter to specify the maximum number of blocks to select for reclustering. This avoids reclustering too many blocks at once.

* Added memory threshold control for recluster selection  

   - Recluster mutator now calculates an in-memory threshold based on system memory and max memory usage setting, and does not select blocks beyond that threshold.

* Optimized merge sort pipelines for recluster and optimize table  

   - Changed full sort to merge sort, and added resize transforms to optimize thread utilization.

* Refactored processors and mutations logs  

   - Renamed and refactored some processors for clearer naming.  
   - Updated mutation logs to use a simpler enum structure.

* Added tests to cover changes
   - Tests for limit parameters of optimize table, purge and clustering information after optimize.

Hope this helps! Let me know if you have any other questions.

**Comment by drmingdrmer** [comment]
> ### PR summary(by Claude):
> This is a summary of the main changes made in the pull request:
> 
> * Added limit parameter to optimize table, purge and recluster operations
>   
>   * optimize table compact now supports a limit parameter to specify the maximum number of blocks to compact. This is useful for controlling the impact on the cluster.
>   * purge now supports a limit parameter to specify the maximum number of snapshot files to purge. This avoids purging too many files at once.
>   * recluster now supports a limit parameter to specify the maximum number of blocks to select for reclustering. This avoids reclustering too many blocks at once.
> * Added memory threshold control for recluster selection
>   
>   * Recluster mutator now calculates an in-memory threshold based on system memory and max memory usage setting, and does not select blocks beyond that threshold.
> * Optimized merge sort pipelines for recluster and optimize table
>   
>   * Changed full sort to merge sort, and added resize transforms to optimize thread utilization.
> * Refactored processors and mutations logs
>   
>   * Renamed and refactored some processors for clearer naming.
>   * Updated mutation logs to use a simpler enum structure.
> * Added tests to cover changes
>   
>   * Tests for limit parameters of optimize table, purge and clustering information after optimize.
> 
> Hope this helps! Let me know if you have any other questions.

@BohuTANG This is built with your prompt 🤔 Looks quite good.

**Review by dantengsky — APPROVED** [review]

**Pull request merged**